### PR TITLE
[FSDP] Allow non-uniform `requires_grad` for `use_orig_params=True`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98250
* #98249
* __->__ #98221

Closes https://github.com/pytorch/pytorch/issues/91167.

Differential Revision: [D44660134](https://our.internmc.facebook.com/intern/diff/D44660134)

### CODE REVIEW

File: torch/distributed/fsdp/flat_param.py [hunk 1]
```diff
@@ -471,7 +471,12 @@ def _init_flat_param_and_metadata(
             raise ValueError(
                 f"Expects non-negative `aligned_numel` but got {aligned_numel}"
             )
-        dtype, requires_grad, device = self._validate_tensors_to_flatten(params)
+        (
+            dtype,
+            requires_grad,
+            device,
+            requires_grad_mask,
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:478]: maybe rename requires_grad and the mask to indicate one is for the entire flat param and the other is per param?

File: torch/distributed/fsdp/flat_param.py [hunk 2]
```diff
@@ -566,6 +571,12 @@ def _init_flat_param_and_metadata(
             _convert_to_params(shared_params) if use_orig_params else None,
             is_padding_mask,
         )
+        if self._use_orig_params:
+            assert self.flat_param._params is not None  # mypy
+            for param, requires_grad in zip(
+                self.flat_param._params, requires_grad_mask
+            ):
+                param.requires_grad = requires_grad
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:579]: why do we need to set this here? If these are the original params, shouldn't this value already be populated?

**Review by rohan-varma — APPROVED** [review]
LG

File: torch/distributed/fsdp/flat_param.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:579]: Good point!

**Comment by rohan-varma** [comment]
@rohan-varma has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44660134).
